### ISSUE TITLE
Permutation group triple product property

### ISSUE THREAD

**Issue by sylee957** [sympy/sympy#18525]
My attempt to see the actual code of coppersmith-winograd algorithm seems like in vain, but I see that they had discovered some useful properties for group theory, like triple product property.
https://en.wikipedia.org/wiki/Triple_product_property

I'd present a naive way of computing here, and see if anyone else here wants to implement this stuff to permutation groups.

```python3
def satisfies_triple_product_property(G, G1, G2, G3):
    """Checks if the permutation groups satisfies the triple product property
    
    References
    ==========
    
    .. [1] https://en.wikipedia.org/wiki/Triple_product_property
    """
    if not G1.is_subgroup(G):
        return False
    if not G2.is_subgroup(G):
        return False
    if not G3.is_subgroup(G):
        return False
    
    G1_elems = set(G1.generate())
    G2_elems = set(G2.generate())
    G3_elems = set(G3.generate())
    
    for g1 in G1_elems:
        for g2 in G2_elems:
            for g3 in G3_elems:
                if (g1*g2*g3).is_identity:
                    if not g1.is_identity:
                        return False
                    if not g2.is_identity:
                        return False
                    if not g3.is_identity:
                        return False
    return True

G = AbelianGroup(2, 2, 2)
G1 = PermutationGroup(Permutation(0, 1, size=6))
G2 = PermutationGroup(Permutation(2, 3, size=6))
G3 = PermutationGroup(Permutation(4, 5, size=6))

satisfies_triple_product_property(G, G1, G2, G3)
```

**Comment by mohitacecode** [comment]
I want to work on this? @sylee957

**Comment by sylee957** [comment]
Yes, but the code may need some refinement.

**Comment by mohitacecode** [comment]
ok I will first give some time to the given link. should I refine this code?

**Comment by sylee957** [comment]
I think that there are some, but not many matherials that can be useful.
https://arxiv.org/pdf/1104.5097.pdf
(It also has the same algorithm that I have made here.)

**Comment by mohitacecode** [comment]
There are total of 5 algorithms given in this article(mentioned by you) and each algorithm has different complexity.
I think that the algorithms that you have implemented is `Algorithm 5.2. TPPTest( S, T, U )`
but in the article the algorithm `Algorithm 5.5 TPPTestMurthy` was found most efficient.

and I am confused are we going to implement TPP for the subsets or subgroup.(article has mentioned algorithms for both)

**Comment by sylee957** [comment]
You can test the algorithm on the generating set of the group, if the algorithm is about subsets.

**Comment by mohitacecode** [comment]
> I think that the algorithms that you have implemented is `Algorithm 5.2. TPPTest( S, T, U )`
> but in the article the algorithm `Algorithm 5.5 TPPTestMurthy` was found most efficient.

What about the algorithm that we should use?

**Comment by sylee957** [comment]
I think that things can be different with sympy, due to the abstractions of how multiplication or such are computed in sympy, so I cannot tell which algorithm would be the best by only reading the benchmarks which is not experimented with sympy.
So you should benchmark each methods carefully with implementation in sympy, if you want to choose one.

But if you don't have an idea how to reasonably test against so many groups on the earth, I'm fine with providing a quantitative API by implementing everything and make a keyword `method='murphy'` to select each methods.

**Comment by mohitacecode** [comment]
ok let's start with first two algorithms we have mentioned.

**Comment by mohitacecode** [comment]
Do we have function in sympy to find intersection of two group?

**Comment by sylee957** [comment]
I don’t think that it exists. You should probably generate the sets manually (by using G.generate()) and compute the intersection.

**Comment by mohitacecode** [comment]
hey I have one doubt is intersection of two group is same as intersection of two sets as far as I know?

**Comment by sylee957** [comment]
Generally, the notation of intersection between two groups, are defined as a intersection between two underlying sets of the group.

**Comment by mohitacecode** [comment]
ok groups are also sets but they satisfy some special properties that make them sets.

**Comment by sylee957** [comment]
I don't think that the paper is abusing the intersection notation. You can see that most of the algorithms there are checking whether the intersection is a trivial group or an empty set.

**Comment by mohitacecode** [comment]
Yeah you seems right they are checking `whether the intersection is a trivial group`.

There's a theoram that says : `Two Normal Subgroups Intersecting Trivially Commute Each Other`
here's a link for more details article : https://yutsumura.com/two-normal-subgroups-intersecting-trivially-commute-each-other/

should we use this?

**Comment by sylee957** [comment]
I can’t certify that whether the normal group property helps solving this problem, because to find whether two normal subgroups commute each other may not be fast than the algorithms presented in the paper

**Comment by mohitacecode** [comment]
But algorithms is for sure using the whether intersection of two subgroup is trivial

```
def test(S, T, U):
    QT := Q(T), QU := Q(U)
    if( QT ∩ QU = 1 ) then
        QS := Q(S)
        if( QS ∩ QTQU = 1 ) then
            return true;
    fi; fi; 
    return false;
```

**Comment by mohitacecode** [comment]
ping @jksuom @sylee957
